File: app/src/main/java/com/simplemobiletools/musicplayer/MyWidgetProvider.java
package com.simplemobiletools.musicplayer;

import android.app.PendingIntent;
import android.appwidget.AppWidgetManager;
import android.appwidget.AppWidgetProvider;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.res.Resources;
import android.graphics.Bitmap;
import android.graphics.Color;
import android.os.Bundle;
import android.widget.RemoteViews;

import com.simplemobiletools.musicplayer.activities.MainActivity;
import com.squareup.otto.Bus;
import com.squareup.otto.Subscribe;

public class MyWidgetProvider extends AppWidgetProvider {
    private static RemoteViews mRemoteViews;
    private static AppWidgetManager mWidgetManager;
    private static Context mContext;
    private static Intent mIntent;
    private static Bus mBus;
    private static Song mCurrSong;
    private static Bitmap mPlayBitmap;
    private static Bitmap mPauseBitmap;

    private static boolean mIsPlaying;
    private static int[] mWidgetIds;

    @Override
    public void onUpdate(Context context, AppWidgetManager appWidgetManager, int[] appWidgetIds) {
        initVariables(context);
        super.onUpdate(context, appWidgetManager, appWidgetIds);
    }

    private void setupIntent(String action, int id) {
        mIntent.setAction(action);
        final PendingIntent pendingIntent = PendingIntent.getBroadcast(mContext, 0, mIntent, 0);

        if (mRemoteViews != null)
            mRemoteViews.setOnClickPendingIntent(id, pendingIntent);
    }

    private void setupAppOpenIntent(int id) {
        final Intent intent = new Intent(mContext, MainActivity.class);
        final PendingIntent pendingIntent = PendingIntent.getActivity(mContext, 0, intent, 0);
        mRemoteViews.setOnClickPendingIntent(id, pendingIntent);
    }

    private void initVariables(Context context) {
        mContext = context;
        mIntent = new Intent(mContext, MyWidgetProvider.class);
        updateWidgetIds();
        for (int widgetId : mWidgetIds) {
            mRemoteViews = getRemoteViews(mWidgetManager, mContext, widgetId);
        }

        setupViews(mContext);

        if (mBus == null) {
            mBus = BusProvider.getInstance();
        }
        registerBus();
    }

    private void updateWidgetIds() {
        final ComponentName component = new ComponentName(mContext, MyWidgetProvider.class);
        mWidgetManager = AppWidgetManager.getInstance(mContext);
        mWidgetIds = mWidgetManager.getAppWidgetIds(component);
    }

    private SharedPreferences initPrefs(Context context) {
        return context.getSharedPreferences(Constants.PREFS, Context.MODE_PRIVATE);
    }

    @Subscribe
    public void songChangedEvent(Events.SongChanged event) {
        mCurrSong = event.getSong();
        updateSongInfo();
        updateWidgets();
    }

    private void updateSongInfo() {
        String title = "";
        String artist = "";
        if (mCurrSong != null) {
            title = mCurrSong.getTitle();
            artist = mCurrSong.getArtist();
        }

        if (mRemoteViews == null)
            return;

        mRemoteViews.setTextViewText(R.id.songTitle, title);
        mRemoteViews.setTextViewText(R.id.songArtist, artist);
    }

    @Subscribe
    public void songStateChanged(Events.SongStateChanged event) {
        if (mIsPlaying == event.getIsPlaying())
            return;

        mIsPlaying = event.getIsPlaying();
        updatePlayPauseButton();
        updateWidgets();
    }

    private void updatePlayPauseButton() {
        Bitmap bmp = mPlayBitmap;

        if (mIsPlaying)
            bmp = mPauseBitmap;

        mRemoteViews.setImageViewBitmap(R.id.playPauseBtn, bmp);
    }

    private void updateWidgets() {
        for (int widgetId : mWidgetIds) {
            mWidgetManager.updateAppWidget(widgetId, mRemoteViews);
        }
    }

    private void updateColors(Context context) {
        final SharedPreferences prefs = initPrefs(context);
        final Resources res = context.getResources();
        final int defaultColor = res.getColor(R.color.dark_grey_transparent);
        final int newBgColor = prefs.getInt(Constants.WIDGET_BG_COLOR, defaultColor);
        final int newTextColor = prefs.getInt(Constants.WIDGET_TEXT_COLOR, Color.WHITE);

        if (mRemoteViews == null)
            initVariables(context);

        mRemoteViews.setInt(R.id.widget_holder, "setBackgroundColor", newBgColor);

        mRemoteViews.setInt(R.id.songTitle, "setTextColor", newTextColor);
        mRemoteViews.setInt(R.id.songArtist, "setTextColor", newTextColor);

        Bitmap bmp = Utils.getColoredIcon(res, newTextColor, R.mipmap.previous);
        mRemoteViews.setImageViewBitmap(R.id.previousBtn, bmp);

        mPlayBitmap = Utils.getColoredIcon(res, newTextColor, R.mipmap.play);
        mPauseBitmap = Utils.getColoredIcon(res, newTextColor, R.mipmap.pause);

        bmp = Utils.getColoredIcon(res, newTextColor, R.mipmap.next);
        mRemoteViews.setImageViewBitmap(R.id.nextBtn, bmp);
    }

    @Override
    public void onReceive(Context context, Intent intent) {
        if (mRemoteViews == null || mWidgetManager == null || mBus == null || mContext == null) {
            initVariables(context);
        }

        final String action = intent.getAction();
        switch (action) {
            case Constants.PREVIOUS:
            case Constants.PLAYPAUSE:
            case Constants.NEXT:
                Utils.sendIntent(context, action);
                break;
            default:
                super.onReceive(context, intent);
        }
    }

    @Override
    public void onDeleted(Context context, int[] appWidgetIds) {
        super.onDeleted(context, appWidgetIds);
        unregisterBus();
        updateWidgetIds();
    }

    private void setupButtons() {
        setupIntent(Constants.PREVIOUS, R.id.previousBtn);
        setupIntent(Constants.PLAYPAUSE, R.id.playPauseBtn);
        setupIntent(Constants.NEXT, R.id.nextBtn);

        setupAppOpenIntent(R.id.songTitle);
        setupAppOpenIntent(R.id.songArtist);
    }

    private void setupViews(Context context) {
        updateWidgetIds();
        updateColors(context);
        setupButtons();
        updateSongInfo();
        updatePlayPauseButton();
        updateWidgets();
    }

    @Override
    public void onAppWidgetOptionsChanged(Context context, AppWidgetManager appWidgetManager, int widgetId, Bundle newOptions) {
        mRemoteViews = getRemoteViews(appWidgetManager, context, widgetId);
        mWidgetManager = appWidgetManager;
        setupViews(context);
        super.onAppWidgetOptionsChanged(context, appWidgetManager, widgetId, newOptions);
    }

    private RemoteViews getRemoteViews(AppWidgetManager appWidgetManager, Context context, int widgetId) {
        final Bundle options = appWidgetManager.getAppWidgetOptions(widgetId);
        final int minHeight = options.getInt(AppWidgetManager.OPTION_APPWIDGET_MIN_HEIGHT);
        final int rows = getCellsForSize(minHeight);

        int layoutId = R.layout.widget;
        if (rows == 1)
            layoutId = R.layout.small_widget;

        return new RemoteViews(context.getPackageName(), layoutId);
    }

    private static int getCellsForSize(int size) {
        int n = 2;
        while (70 * n - 30 < size) {
            ++n;
        }
        return n - 1;
    }

    private void registerBus() {
        try {
            mBus.register(this);
        } catch (Exception e) {
        }
    }

    private void unregisterBus() {
        try {
            mBus.unregister(this);
        } catch (Exception e) {
        }
    }
}

File: app/src/main/kotlin/com/simplemobiletools/musicplayer/helpers/MyWidgetProvider.kt
package com.simplemobiletools.musicplayer.helpers

import android.app.PendingIntent
import android.appwidget.AppWidgetManager
import android.appwidget.AppWidgetProvider
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.Intent.FLAG_ACTIVITY_NEW_TASK
import android.os.Bundle
import android.widget.RemoteViews
import androidx.media3.common.MediaMetadata
import androidx.media3.session.MediaController
import com.simplemobiletools.commons.extensions.applyColorFilter
import com.simplemobiletools.commons.extensions.getColoredBitmap
import com.simplemobiletools.commons.extensions.getLaunchIntent
import com.simplemobiletools.musicplayer.R
import com.simplemobiletools.musicplayer.activities.SplashActivity
import com.simplemobiletools.musicplayer.extensions.config
import com.simplemobiletools.musicplayer.extensions.maybePreparePlayer
import com.simplemobiletools.musicplayer.extensions.togglePlayback
import com.simplemobiletools.musicplayer.playback.PlaybackService

class MyWidgetProvider : AppWidgetProvider() {

    override fun onUpdate(context: Context, appWidgetManager: AppWidgetManager, appWidgetIds: IntArray) {
        performUpdate(context)
    }

    private fun performUpdate(context: Context) {
        val appWidgetManager = AppWidgetManager.getInstance(context) ?: return
        appWidgetManager.getAppWidgetIds(getComponentName(context)).forEach {
            val views = getRemoteViews(appWidgetManager, context, it)
            updateColors(context, views)
            setupButtons(context, views)
            updateSongInfo(views, PlaybackService.currentMediaItem?.mediaMetadata)
            updatePlayPauseButton(context, views, PlaybackService.isPlaying)
            appWidgetManager.updateAppWidget(it, views)
        }
    }

    override fun onEnabled(context: Context) = triggerUpdate(context)

    override fun onReceive(context: Context, intent: Intent) {
        when (val action = intent.action) {
            TRACK_STATE_CHANGED -> performUpdate(context)
            PREVIOUS, PLAYPAUSE, NEXT -> handlePlayerControls(context, action)
            else -> super.onReceive(context, intent)
        }
    }

    override fun onAppWidgetOptionsChanged(context: Context, appWidgetManager: AppWidgetManager, widgetId: Int, newOptions: Bundle) = triggerUpdate(context)

    private fun handlePlayerControls(context: Context, action: String) {
        maybePreparePlayer(context) { player, _ ->
            if (player.currentMediaItem == null) {
                val intent = context.getLaunchIntent() ?: Intent(context, SplashActivity::class.java)
                intent.addFlags(FLAG_ACTIVITY_NEW_TASK)
                context.startActivity(intent)
            } else {
                when (action) {
                    NEXT -> player.seekToNextMediaItem()
                    PREVIOUS -> player.seekToPreviousMediaItem()
                    PLAYPAUSE -> player.togglePlayback()
                }
            }
        }
    }

    private fun maybePreparePlayer(context: Context, callback: (player: MediaController, prepared: Boolean) -> Unit) {
        SimpleMediaController.getInstance(context).withController {
            maybePreparePlayer(context) { success ->
                callback(this, success)
            }
        }
    }

    private fun triggerUpdate(context: Context) {
        performUpdate(context)
        maybePreparePlayer(context) { _, success ->
            if (success) {
                performUpdate(context)
            }
        }
    }

    private fun setupIntent(context: Context, views: RemoteViews, action: String, id: Int) {
        val intent = Intent(context, MyWidgetProvider::class.java)
        intent.action = action
        val pendingIntent = PendingIntent.getBroadcast(context, 0, intent, PendingIntent.FLAG_IMMUTABLE)
        views.setOnClickPendingIntent(id, pendingIntent)
    }

    private fun setupAppOpenIntent(context: Context, views: RemoteViews, id: Int) {
        val intent = context.getLaunchIntent() ?: Intent(context, SplashActivity::class.java)
        val pendingIntent = PendingIntent.getActivity(context, 0, intent, PendingIntent.FLAG_IMMUTABLE)
        views.setOnClickPendingIntent(id, pendingIntent)
    }

    private fun updateSongInfo(views: RemoteViews, currSong: MediaMetadata?) {
        if (currSong != null) {
            views.setTextViewText(R.id.song_info_title, currSong.title)
            views.setTextViewText(R.id.song_info_artist, currSong.artist)
        }
    }

    private fun updatePlayPauseButton(context: Context, views: RemoteViews, isPlaying: Boolean) {
        val drawableId = if (isPlaying) com.simplemobiletools.commons.R.drawable.ic_pause_vector else com.simplemobiletools.commons.R.drawable.ic_play_vector
        val widgetTextColor = context.config.widgetTextColor
        val icon = context.resources.getColoredBitmap(drawableId, widgetTextColor)
        views.setImageViewBitmap(R.id.play_pause_btn, icon)
    }

    private fun updateColors(context: Context, views: RemoteViews) {
        val config = context.config
        val widgetBgColor = config.widgetBgColor
        val widgetTextColor = config.widgetTextColor

        views.apply {
            applyColorFilter(R.id.widget_background, widgetBgColor)
            setTextColor(R.id.song_info_title, widgetTextColor)
            setTextColor(R.id.song_info_artist, widgetTextColor)
            setImageViewBitmap(
                R.id.previous_btn,
                context.resources.getColoredBitmap(
                    com.simplemobiletools.commons.R.drawable.ic_previous_vector,
                    widgetTextColor
                )
            )
            setImageViewBitmap(R.id.next_btn, context.resources.getColoredBitmap(com.simplemobiletools.commons.R.drawable.ic_next_vector, widgetTextColor))
        }
    }

    private fun setupButtons(context: Context, views: RemoteViews) {
        setupIntent(context, views, PREVIOUS, R.id.previous_btn)
        setupIntent(context, views, PLAYPAUSE, R.id.play_pause_btn)
        setupIntent(context, views, NEXT, R.id.next_btn)

        setupAppOpenIntent(context, views, R.id.song_info_title)
        setupAppOpenIntent(context, views, R.id.song_info_artist)
    }

    private fun getRemoteViews(appWidgetManager: AppWidgetManager, context: Context, widgetId: Int): RemoteViews {
        val options = appWidgetManager.getAppWidgetOptions(widgetId)
        val minHeight = options.getInt(AppWidgetManager.OPTION_APPWIDGET_MIN_HEIGHT)

        if (widgetId == context.config.widgetIdToMeasure && context.config.initialWidgetHeight == 0) {
            context.config.initialWidgetHeight = minHeight
        }

        val layoutId = if (minHeight < context.config.initialWidgetHeight / 2) {
            R.layout.small_widget
        } else {
            R.layout.widget
        }

        return RemoteViews(context.packageName, layoutId)
    }

    private fun getComponentName(context: Context) = ComponentName(context, MyWidgetProvider::class.java)
}
